### DISCUSSION TITLE
auto converge feature request! could be a game changer!

### DISCUSSION THREAD

**Discussion by donrikk** [nagadomi/nunif#83] [Ideas]
HI! I'm from the discord that was created and I've been tinkering with this and other programs and I find that the program 3Dcombine has this feature called auto converge. This auto converge feature has 2 options, converge around screen which is basically what your program would render out as. and then there is the option to auto converge OUT of screen. 

owl3D uses a pipeline called temporal coherence which tracks foreground object in a given scene, owl uses this to help fix edges around foreground objects BUT I have a possible other use case for this!

you could take this concept of temporal coherence and set it to track foreground like owl3D does BUT instead of implementing it to fix edges you could implement it to keep track for the foreground object and whenever ta object or person comes into the foreground the video will adjust those objects or people in real time frame by frame to increase the pop out effect for scenes focusing on foreground shots, then when it goes back to a wider shot where nothing is being directly picked up in the foreground it'll go back to normal depth and convergence until the temporal coherence is called upon again for another foreground scene. 

now I wont pretend to know if something like this is easily implementable but I'm pretty sure u could definitely use this temporal coherence pipeline in my proposed work flow. please let me know what u think of this and if this is something u could do.

honestly this is the biggest thing setting a lot of 2d to 3d converters back is the lack of auto converge capability. all 3D movies and its depths per scene are all dynamically different, a wide shot would look different from a close up shot in theaters the convergence would be manipulated frame by frame in  the same manner which is how pop out technically works, u can protrude the image out of screen all u want but if u don't have the visual convergence to accompany it then it'll never really perform the way someone would like it too.

in fact u might even be able to use the background removal system you have right now in that manner im not sure how u would but if it segments the foreground from the background that is another way of tracking foreground objects aswell or at least could be manipulated that way.

**Comment by donrikk** [comment]
you also have the choose your own size feature which would already help with image stretching as it adds a slight pad to the image allowing room forthe images to shift left and right without image distortion. if you were to implement this auto converge feature i think ur tools will already pretty much work ahnd and hand with it.

**Comment by loawizard** [comment]
YES!! That would be great!

**Comment by nagadomi** [comment]
I'm not really sure what you are looking for.
What is the aim of `Auto converge` operation?

3DCombine manual describes the following.
>Auto Converge - left and right image are moved closer together or further apart (to do this manually, use the Tweak menu) for better stereo effect. The number of pixels shifted is indicated. It aligns the images more correctly, avoiding eye strain in watching them particuarly in head sets. This step is taken after the 3D image is created.

iw3's output is generated from a single image, so there is no misalignment between the left and right cameras which means that calibration is not required.
I tried applying `Auto Converge` to the iw3 output. Then only the image for the right eye was shifted and a more eye strain image was output.

**Reply by donrikk** [reply]
to fix that issue with the right eye, you could try adding 1 pixel of horizontal padding that way when it auto converges the padding will be recognized as the edge so it will stretch and align from the black bar instead of the image, right now the image is being warped on the left or right side of the image when a lot of convergence is added but in 3D bluray rips this isnt present because they use padding so it adjust convergence from the black bars not hte image itself. if you watch this video from the link i will clip to this, if u watch closely to the sides of the sbs video. youll notice hte bars are moving in and out, this is a example of auto convergence. so when a wide shot is happening it kinda stays at that default convergence then when something enters the foreground those bars adjust according ly to whre the convergence needs to be to center onto the foreground image, the current convergence used in your program is base convergence for adding extra depth to your image which is great and needed, but then there has to be a system in place that converges and diverges that convergence so say a foreground object is coming into line of sight, the aut oconverge would start moving closer to 1 convergence and when the subject leaves the foreground it would automatically go back to 0 convergence.[https://www.youtube.com/watch?v=GjgfH_Xgxog](url)

**Reply by donrikk** [reply]
and i think the reason why the right side isnt converging is because it doesn't have too. so usually in a studio conversion they only do left side depth map renders not both sides like this program does, so since its already taken into account the left eye depth information it doesnt need to converge it only takes the right eye to converge since the left eye is already converged where it needs to be while the right eye needs to be manipulated.

**Reply by donrikk** [reply]
so basically the convergence level should be from 0 to 1, so during shots with no viewable foreground object it'll be consistently at 0 convergence then auto converge would kick in when it sees a object start to enter the foreground and start adjusting convergence the closer it got to the foreground until it finally reaches its max convergence of 1. what this does is mimics the human eye, when something is very close to you and u attempt to look at it your eyes converge onto that object thus making it the center of your attention. but when someone isn't closely in front of you then u get a full wide view of your surroundings with your eyes base convergence.

basically u want to make the video interact with the user the way your eyes would interact with real life situations and one thing eyes are constantly doing is converging.

**Reply by donrikk** [reply]
if u want to add a smart pad option and set your horizontal in that setting one above ur current resolution so if ur horizontal is 1920 ude make it 1921, then keep vertical the same and u wont see it with ur naked eye but there is a black border around the entire video in combine now. then u can add ur auto convergence and it will converge from the black borders not from the image itself.

**Reply by nagadomi** [reply]
I think I got it.
Maybe this can be implemented in various methods, from very simple to very complicated.
I would try the easiest method after finishing other TODOs.

**Reply by donrikk** [reply]
sounds tantastic! im excited to see waht u guys come up with!

**Reply by loawizard** [reply]
AWESOME!

**Reply by kaelsonofkrypto** [reply]
This would totally be an excellent feature to add.

**Comment by loawizard** [comment]
This would make IW3 Next level
